src/common/meta/src/wal/options_allocator.rs

```diff
@@ -34,37 +34,62 @@ pub enum WalOptionsAllocator {
 
 impl WalOptionsAllocator {
     /// Creates a WalOptionsAllocator.
-    pub fn new(config: &WalConfig, kv_backend: KvBackendRef) -> Self {
-        todo!()
+    pub fn new(config: WalConfig, kv_backend: KvBackendRef) -> Self {
+        match config {
+            WalConfig::RaftEngine => Self::RaftEngine,
+            WalConfig::Kafka(kafka_config) => {
+                Self::Kafka(KafkaTopicManager::new(kafka_config, kv_backend))
+            }
+        }
     }
 
     /// Tries to initialize the allocator.
-    pub fn try_init(&self) -> Result<()> {
-        todo!()
+    pub async fn try_init(&mut self) -> Result<()> {
+        match self {
+            Self::RaftEngine => Ok(()),
+            Self::Kafka(kafka_topic_manager) => kafka_topic_manager.try_init().await,
+        }
     }
 
     /// Allocates a wal options for a region.
     pub fn alloc(&self) -> WalOptions {
-        todo!()
+        match self {
+            Self::RaftEngine => WalOptions::RaftEngine,
+            Self::Kafka(kafka_topic_manager) => WalOptions::Kafka(KafkaWalOptions {
+                topic: kafka_topic_manager.select().clone(),
+            }),
+        }
     }
 
     /// Allocates a batch of wal options where each wal options goes to a region.
     pub fn alloc_batch(&self, num_regions: usize) -> Vec<WalOptions> {
         match self {
             WalOptionsAllocator::RaftEngine => vec![WalOptions::RaftEngine; num_regions],
-            WalOptionsAllocator::Kafka(topic_manager) => {
-                let topics = topic_manager.select_batch(num_regions);
-                topics
-                    .into_iter()
-                    .map(|topic| WalOptions::Kafka(KafkaWalOptions { topic }))
-                    .collect()
-            }
+            WalOptionsAllocator::Kafka(topic_manager) => topic_manager
+                .select_batch(num_regions)
+                .into_iter()
+                .map(|topic| {
+                    WalOptions::Kafka(KafkaWalOptions {
+                        topic: topic.clone(),
+                    })
+                })
+                .collect(),
         }
     }
 }
 
+/// Creates and initializes a wal options allocator.
+pub async fn build_wal_options_allocator(
+    config: &WalConfig,
+    kv_backend: &KvBackendRef,
+) -> Result<WalOptionsAllocator> {
+    let mut allocator = WalOptionsAllocator::new(config.clone(), kv_backend.clone());
+    allocator.try_init().await?;
```

Comment:
I think the `try_init` must be called outside of this building function. It's better to keep functions orthogonal, one function only tries to do one thing.

Reply:
Another suggestion: we often use `start` for such initialization procedures.

Reply:
Can we initialize the allocator in `new()`?

Reply:
I prefer to do the initialization in `new` or `try_new`. But some comments suggest to not mix `new` and `start`. The advantage is that the method name implies that a heavy operation may be executed by the `start` method.

Reply:
The `try_init` method is renamed to `start` and it's called in `MetaStateHandler::on_become_leader`.